wgpu/src/api/queue.rs

```diff
@@ -98,67 +98,82 @@ impl Queue {
-    /// method avoids an intermediate copy and is often able to transfer data
-    /// more efficiently than this one.
+    ///   However, `data` will be immediately copied into staging memory, so the
+    ///   caller may discard it any time after this call completes.
     ///
-    /// Currently on native platforms, for both of these methods the staging
-    /// memory will be a new allocation. This will then be released after the
-    /// next submission finishes. To entirely avoid short-lived allocations, you might
-    /// be able to use [`StagingBelt`](crate::util::StagingBelt).
+    /// * Consider using [`Queue::write_buffer_with()`] instead.
+    ///   That method allows you to prepare your data directly within the staging
+    ///   memory, rather than first placing it in a separate `[u8]` to be copied.
+    ///   That is, `queue.write_buffer(b, offset, data)` is approximately equivalent
+    ///   to `queue.write_buffer_with(b, offset, data.len()).copy_from_slice(data)`,
+    ///   so use `write_buffer_with()` if you can do something smarter than that
+    ///   [`copy_from_slice()`](slice::copy_from_slice). However, for small values
+    ///   (e.g. a typical uniform buffer whose contents come from a `struct`),
+    ///   there will likely be no difference, since the compiler will be able to
+    ///   optimize out unnecessary copies regardless.
+    ///
+    /// * Currently on native platforms, for both of these methods, the staging
+    ///   memory will be a new allocation. This will then be released after the
+    ///   next submission finishes. To entirely avoid short-lived allocations, you might
+    ///   be able to use [`StagingBelt`](crate::util::StagingBelt),
+    ///   or buffers you explicitly create, map, and unmap yourself.
     pub fn write_buffer(&self, buffer: &Buffer, offset: BufferAddress, data: &[u8]) {
         self.inner.write_buffer(&buffer.inner, offset, data);
     }
 
-    /// Write to a buffer via a directly mapped staging buffer.
+    /// Prepares to write data to a buffer via a mapped staging buffer.
+    ///
+    /// This operation allocates a temporary buffer and then returns a
+    /// [`QueueWriteBufferView`], which
+    ///
+    /// * dereferences to a `[u8]` of length `size`, and
+    /// * when dropped, schedules a copy of its contents into `buffer` at `offset`.
     ///
-    /// Return a [`QueueWriteBufferView`] which, when dropped, schedules a copy
-    /// of its contents into `buffer` at `offset`. The returned view
-    /// dereferences to a `size`-byte long `&mut [u8]`, in which you should
-    /// store the data you would like written to `buffer`.
+    /// Therefore, this obtains the same result as [`Queue::write_buffer()`], but may
+    /// allow you to skip one allocation and one copy of your data, if you are able to
+    /// assemble your data directly into the returned [`QueueWriteBufferView`] instead of
+    /// into a separate allocation like a [`Vec`](alloc::vec::Vec) first.
```

Comment:
I think `write_buffer_with()` could really use a code example demonstrating what this sort of use, but it’ll require some further thought to come up with a nice short one, so this PR doesn’t do that. (Also, it'll be messy no matter what because there are no current guarantees about what *CPU-side alignment* a mapped buffer view has (see #3508), so you have to use byte-by-byte or explicitly unaligned writes.)